### ISSUE TITLE
How do I provide a custom error boundary for my entry app's entry point (/app) and child routes (app/foo, app/foo/baz)?

### ISSUE THREAD

**Issue by muchisx** [Shopify/shopify-app-template-remix#597]
https://github.com/Shopify/shopify-app-template-remix/blob/b7963283746d9a25afd1444c4cdfeb132984b08d/app/routes/app.tsx#L32-L35

This is very confusing for me and I haven't gotten around to understand can I also return a component in my routes that also don't mess up what you guys intend the error boundary to do, which is to use that boundary import from shopify.

I went [here in the docs](https://shopify.dev/docs/api/shopify-app-remix#boundaries) but I left even more confused. I did not grasp how to do it and there's no examples.

I'm asking because currently without error boundaries my app just brakes whenever I have an error server-side and I have to restart my local server every time.

If someone could explain me like I'm 5 haha, would be great, thanks !

**Comment by paulomarg** [comment]
Hey! I hear you, error boundaries are tricky right now - we're looking at ways to improve them, but for now we need to do that.

We need that error boundary to be in `app.tsx` because there are flows where we need to throw responses with headers, so we need to catch those throws at this level for the `headers` function to kick in.

`boundary.error` will do 2 things ([code for context](https://github.com/Shopify/shopify-app-js/blob/main/packages/shopify-app-remix/src/server/boundary/error.tsx)):
- If a `Response` was thrown, we'll return some HTML to render the page normally so the headers are included. This behaviour needs to stay, or the app won't authenticate properly.
- If an `Error` was thrown, we'll just rethrow it. As far as our code goes, we don't care what happens after this.

All that aside, if you just want to add some custom code when errors happen, you can do this:

```ts
export const ErrorBoundary = () => {
  const error = useRouteError();
  try {
    // Catch redirect Response thrown by @shopify/shopify-app-remix
    const shopifyError = boundary.error(error);
    return shopifyError;
  } catch {
    // Run your own code here
    return <p>Something went wrong ☹️</p>;
  }
};
```

We'll see if we can add some documentation on how to add your own boundaries, but it will basically be the above :)

Hope this helps!

**Comment by nboliver-ventureweb** [comment]
@paulomarg What is the recommended to process to throw a 404 response from a loader, as per the Remix docs? https://remix.run/docs/hi/main/guides/not-found#how-to-send-a-404

**Comment by umutnacak** [comment]
@nboliver-ventureweb I thought using something similar but after looking at the shopify-app-remix code I understand that the boundary.error function is used for the [bounce page implementation](https://shopify.dev/docs/apps/build/authentication-authorization/set-embedded-app-authorization?extension=javascript#bounce-page-to-get-a-session-token-from-app-bridge). They return a script tag which handles refreshing the session token. So I ended up with something like this:

I send the error like this:
```ts
throw data({ id: "my-unique-id", payload: "my-payload" }, {
  status: 404,
  statusText: "Not Found",
});
```
And catch it like this:
```tsx
if (isRouteErrorResponse(error) && error.data && error.data.id !== "my-unique-id") {
  return (
    <div
      dangerouslySetInnerHTML={{ __html: error.data }}
    />
  );
}
else {
  return <div>Known error or no data handler</div>
}
```
It's not perfect because I had to use their implementation for any other data to make sure that I don't break anything (if not already). But I think this is an upgrade over the infamous 'Handling response' message.

Basically they need to do this check inside the errorBoundary function of the shopify-app-remix package and re-throw instead of showing the 'Handling response' message.

And as for the internal not found error you can use the ErrorBoundary of the root.tsx like shown in this documentation:
https://remix.run/docs/en/main/file-conventions/root#layout-export which shouldn't require the usage of the boundary functions, but I'm not that sure.

**Comment by umutnacak** [comment]
Yes probably boundary.error function should be included for the compliance.
I think `error.data` could be null so it should be `error.data?.id`.

I still think this check should be part of the shopify-app-remix package, but any official suggestion from the developers would be welcomed. I think @lizkenyon is the new maintainer because I see their name all over the place.

**Comment by madfcat** [comment]
Yes, please. Any insights are really welcome. Maybe expand the documentation as well.
